### ISSUE TITLE
Difference in decoder model from original paper?

### ISSUE THREAD

**Issue by adamconkey** [Kaixhin/PlaNet#27]
I have 2 questions about the decoder model:

1. Looking at the original TensorFlow implementation and this one, it seems the decoder models are different. The [TF version](https://github.com/google-research/planet/blob/cbe77fc011299becf6c3805d6007c5bf58012f87/planet/networks/conv_ha.py#L41) appears to decode to a distribution of observations while [this repo](https://github.com/Kaixhin/PlaNet/blob/master/models.py#L87) uses a standard decoder. Is there a reason you chose to use this decoder, and should that result in any performance differences from the original paper? Or maybe is there something I'm missing that there actually isn't this difference between the versions? 
2. The visual decoder in this repo is a function of both the `belief` and `state` and they get concatenated together. In Google's version their decoder is just a function of `state` though. I know in the paper they formally write the observation model as being a function of both, but the state is already a function of the belief, so maybe the dependence is meant to be implicit? I'm curious if you have insights about which is correct - I find the Google code incomprehensible so it's possible I misinterpret what they're doing.

Thanks for your help.

**Comment by Kaixhin** [comment]
1. Optimising the log likelihood of an observation with respect to a Normal distribution with a given mean and a standard deviation of 1 is equivalent to optimising the MSE between the observation and the mean, so they're the same apart from how you interpret them.
2. I based my code on the paper, referring to the code if I had any issues. I would argue that this is the better way to do it, as otherwise the `state` would have to retain all the relevant information from the `belief`, when the `belief` is the best way to retain all deterministic information.

**Comment by adamconkey** [comment]
Thanks for your response, both points make sense. 

I'm having some trouble trying to use this on my own dataset where I have a 7-DOF robot arm picking and placing blocks. I trained on a static dataset of "expert" demonstrations and then take the model offline and use ground-truth action sequences to feed through the model and see what the decoded observations look like. I'm finding the motion of the robot arm look quite accurate, but it has a very hard time localizing the objects on the table. The objects just kind of ghostly pop up all over the table until the robot arm is closer to the object, at which point the object becomes clearer. I guess this is because at that point it starts to have a clearer association with the robot's actions, while when the arm is further away from the object it's a part of the scene that was randomized so much in training it doesn't know where to localize them. I thought maybe the encoder/decoder model could be an issue but maybe it's how I'm training the model? I'm not quite sure now.

Did you ever try running on an environment where there is more than just an agent in the scene? That is, there are additionally objects in the scene with positions that are randomized during training and testing?

**Comment by Kaixhin** [comment]
Pixel-level losses have always been a bit iffy for scenes with small objects. There's a couple of options, from using a higher res input, to a more powerful encoder/decoder, to a different observation model, and possibly even adding the overshooting loss back in.

I have run it on a robot reaching task, which is visually more simple, and didn't have any issues with the target object appearing. Didn't get it to succeed, but I believe the credit assignment for the problem was hard to learn from scratch.

**Comment by adamconkey** [comment]
I appreciate your suggestions, seems we were thinking along the same lines as I'm currently training with a beefed up encoder/decoder, and was planning to double the image resolution if that still wasn't effective. I also am ultimately going to be using a more complicated sensory space with at least depth images, and possibly other things like joint/object state and segmentations. I was thinking that may also help if the representation has to be used to jointly predict these various observations that need to be consistent, albeit a more complicated learning problem.
